### PR TITLE
Yuya/add checkpoints section

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

Add NeMo Dev docs to explain NeMo checkpoints, .nemo .ckpt, regular and distributed.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [x] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: docs/source/checkpoints/dist_ckpt.rst [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+This function is simply a composition of `load_tensors_metadata` and `save`.
+
+Save and Load Strategies
+========================
+There are multiple ways to save a sharded state dict into a serialized checkpoint. They can be provided by the user as saving and loading strategies.
+
+There are four types of strategies:
+1. Saving strategy for ShardedTensors
+#. Saving strategy for "common" data
+#. Loading strategy for ShardedTensors
```
Comment (cuichenx) [inline docs/source/checkpoints/dist_ckpt.rst:242]: this is not rendered correctly

File: docs/source/checkpoints/dist_ckpt.rst [hunk 2]
```diff
@@ -0,0 +1,416 @@
...
+Flattening
+----------
+The `flattened_range` attribute declares that `ShardedTensor.data` represents a slice of a flattened model parameter.
+This corresponds to a transformation used in Distributed Optimizers which flattens the data and shards it along the data-parallel domain.
+
+Extra flattening comes with an efficiency challenge during checkpoint resharding.
+Since flattening is applied after the global tensors is sharded into the grid of local chunks, loading after resharding requires accessing incontiguous data fragments.
+An example solution for that is implemented in the `resharding <https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/dist_checkpointing/strategies/resharding.py>`_ module and involves saving the flattened tensor with a different global shape than the original one.
+
+Example: For a global tensor [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11]] with sharding by TP (tensor-parallel) over the second axis, here are the local shards if TP=2:
```
Comment (cuichenx) [inline docs/source/checkpoints/dist_ckpt.rst:339]: great example here!

File: docs/source/checkpoints/dist_ckpt.rst [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+ShardedObject
+^^^^^^^^^^^^^
+Sometimes there is a need to save arbitrary objects across the ranks.
+ShardedObject allows to structure those objects into arrays of objects with a fixed ``global_shape`` and save/load parts of the arrays on specific ranks.
+
+ShardedTensorFactory
+^^^^^^^^^^^^^^^^^^^^
+The ShardedTensorFactory class defers tensors transformations until they are actually saved.
+A factory can expand a tensor into an arbitrary sub state dict (including all supported entities listed above).
+The need for such deferral will be explained in the ``Tensors transformations``_ section.
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:169]: Link doesn't work. Need single backticks, like the link inside the ShardedTensor section

File: docs/source/checkpoints/dist_ckpt.rst [hunk 4]
```diff
@@ -0,0 +1,421 @@
...
+    assert isinstance(loaded_state_dict['different-key'], torch.Tensor)
+
+    # The key in the state dict is important only from the subsequent `model.load_state_dict`
+    # that usually happens after `dist_checkpointing.load` - the state dict must have
+    # the structure and keys corresponding to the model structure and submodule names.
+    model.load_state_dict(loaded_state_dict)
+
+ShardedTensor
+^^^^^^^^^^^^^
+ShardedTensor is the primary use case for distributed checkpointing - tensor sharding.
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:156]: Technically should have backticks around ShardedTensor and all the other object names?

File: docs/source/checkpoints/dist_ckpt.rst [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+When the sharded state dict is provided as input, it is processed in the following way (see also ``load`` function `documentation <https://docs.nvidia.com/megatron-core/developer-guide/latest/api-guide/dist_checkpointing.html#module-core.dist_checkpointing.serialization>`_):
+
+1. The "common" state dict is loaded from the checkpoint. This forms the base of the resulting state dict.
+2. The ShardedTensorFactories from the input sharded state dict are applied.
+3. The LocalNonPersistentObjects are extracted from the input sharded state dict, unwrapped and added to the resulting state dict.
+4. The ShardedObjects are extracted and loaded from the checkpoint into the resulting state dict.
+5. The ShardedTensors are extracted and loaded from the checkpoint into the resulting state dict.
+6. Factory merges are applied (see `Optimizers`_ for explanation).
+
+This results in a *regular* state dict with plain tensors that can be further processed by the application (which usually means running ``model.load_state_dict(state_dict)``.
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:220]: unmatched parentheses -- need to add closing parentheses before period

File: docs/source/checkpoints/dist_ckpt.rst [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+The result is a sharded state dict with trivial sharding (every tensor is sharded into one big shard).
+
+dist_checkpointing.load_plain_tensors
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+The ``dist_checkpointing.load_plain_tensors`` function is an entry point that allows reading sharded tensors stored in the checkpoint without any sharding (as plain tensors).
+This function is simply a composition of ``load_tensors_metadata`` and ``save``.
+
+Save and Load Strategies
+------------------------
+There are multiple ways to save a sharded state dict into a serialized checkpoint. They can be provided by the user as saving and loading strategies.
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:240]: "They can be provided by the user as saving and loading strategies." this sentence is confusing. You mean something like the objects need to be specified by the user?

File: docs/source/checkpoints/dist_ckpt.rst [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+1. Saving strategy for ShardedTensors
+2. Saving strategy for "common" data
+3. Loading strategy for ShardedTensors
+4. Loading strategy for "common" data
+
+Additionally, ShardedObjects are handled with either "sharded" or "common" strategy depending on its capabilities (``can_handle_sharded_objects`` property).
+
+Each saving strategy is associated with a ``backend`` and a ``version``.
+Each loading strategy can be associated with multiple ``backend``s and ``version``s it can load.
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:252]: text inside backticks doesn't render correctly due to "s" after it.

Maybe:
"Each loading strategy can be associated with multiple values of ``backend`` and ``version``." ?

Comment (erastorgueva-nv) [reply docs/source/checkpoints/dist_ckpt.rst:252]: or just don't use backticks in this sentence (since you don't use them in the next sentence anyway. or instead, maybe should use backticks in all sentences)

File: docs/source/checkpoints/dist_ckpt.rst [hunk 8]
```diff
@@ -0,0 +1,421 @@
...
+Strategies are the main way to introduce optimizations to the saving and loading algorithm without altering the checkpoint format.
+
+In the following example, the "fully parallel" wrappers modify the saving and loading *algorithm*, but the underlying checkpoint *format* (and ``backend`` in consequence) stays the same.
+It makes the ``basic_save_load`` and ``fully_parallel_save_load`` functions equivalent:
+
+.. code-block:: python
+
+    from megatron.core import dist_checkpointing
+    from megatron.core.dist_checkpointing.strategies.torch import TorchDistLoadShardedStrategy, TorchDistSaveShardedStrategy
+    from megatron.core.dist_checkpointing.strategies.fully_parallel import FullyParallelLoadStrategyWrapper, FullyParallelSaveStrategyWrapper
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:263]: maybe use parentheses here to make the lines shorter. otherwise they go out of view. similarly some of the lines in `full_parallel_save_load` are too long

File: docs/source/checkpoints/dist_ckpt.rst [hunk 9]
```diff
@@ -0,0 +1,421 @@
...
+        assert fully_parallel_save_strategy.backend == base_save_strategy.backend == 'torch_dist'
+        fully_parallel_load_strategy = FullyParallelLoadStrategyWrapper(base_load_strategy)
+        dist_checkpointing.save(sharded_state_dict, ckpt_dir, fully_parallel_save_strategy)
+        return dist_checkpointing.load(sharded_state_dict, ckpt_dir, fully_parallel_load_strategy)
+
+
+The ``dist_checkpointing`` package provides default strategies for some sharded backends, so it's enough to specify a tuple ``(backend, version)`` as a saving strategy.
+Backends and versions are stored in a ``metadata.json`` file inside the checkpoint so that the loading strategy can be determined automatically (provided that there exists a default loading strategy for a given backend and version).
+
+For "sharded" strategies, currently the backends supported by default are based on `PyTorch Distributed`_ format (``torch_dist`` backend) and `Zarr`_ format (``zarr`` backend).
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:288]: currently, if you click on "Zarr" here, it will take you to https://zarr.readthedocs.io/en/stable/.
We want to go to the Zarr section below though, right? To make the docs behave like this, we need to change the line in the Zarr section so that it uses 2 underscores after the hyperlink and backtick (ie `` `Zarr <https://zarr.readthedocs.io/en/stable/>`__  ``

File: docs/source/checkpoints/dist_ckpt.rst [hunk 10]
```diff
@@ -0,0 +1,421 @@
...
+
+Tensors Transformations
+-----------------------
+The ShardedTensor API enables the declaration of basic transformations that should be performed during saving and loading.
+
+Shape Mismatch
+^^^^^^^^^^^^^^
+The ``allow_shape_mismatch`` flag relaxes the requirement of matching global tensor shapes during loading.
+Extra padding is filled with zeros or stripped depending on the mismatch kind.
+This comes handy for layers like embedding which might be padded according to parallelism for performance reasons.
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:333]: correct grammar would be "This comes in handy". Let's make it less informal though: "This is useful"

File: docs/source/checkpoints/dist_ckpt.rst [hunk 11]
```diff
@@ -0,0 +1,421 @@
...
+Flattening
+^^^^^^^^^^
+The ``flattened_range`` attribute declares that ``ShardedTensor.data`` represents a slice of a flattened model parameter.
+This corresponds to a transformation used in Distributed Optimizers which flattens the data and shards it along the data-parallel domain.
+
+Extra flattening comes with an efficiency challenge during checkpoint resharding.
+Since flattening is applied after the global tensors is sharded into the grid of local chunks, loading after resharding requires accessing incontiguous data fragments.
+An example solution for that is implemented in the `resharding <https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/core/dist_checkpointing/strategies/resharding.py>`_ module and involves saving the flattened tensor with a different global shape than the original one.
+
+Example: For a global tensor [[0, 1, 2, 3, 4, 5], [6, 7, 8, 9, 10, 11]] with sharding by TP (tensor-parallel) over the second axis, here are the local shards if TP=2:
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:344]: wrap tensor in backticks?

File: docs/source/checkpoints/dist_ckpt.rst [hunk 12]
```diff
@@ -0,0 +1,421 @@
...
+
+
+Arbitrary Transformations
+^^^^^^^^^^^^^^^^^^^^^^^^^
+The way to apply arbitrary transformations to the tensors during saving and loading is with ShardedTensorFactory.
+It defines such transformations as a function that can be reapplied to any ShardedTensor (in particular, a ShardedTensor representing optimizer states).
+Such "build" function is also tied to a "merge" function that can apply an inverse transformation during loading.
+
+If handling an optimizer state is not required, such transformation could be also applied directly during sharded state dict creation.
+In order to apply such transformation both to model and optimizer parameters in a consistent manner, it's necessary to encode them as factory functions (with original model parameter as the ``data`` input so that the optimizer params can be properly mapped to model ShardedTensors).
```
Comment (erastorgueva-nv) [inline docs/source/checkpoints/dist_ckpt.rst:408]: "such a tranformation"

**Review by erastorgueva-nv — APPROVED** [review]
Looks good, have suggested some small changes.

**Extra note:**

In `user_guide.rst`, maybe prepend all script paths with NeMo/scripts/checkpoint_converters ?

**Review by erastorgueva-nv — APPROVED** [review]
LGTM now, thank you
